tools/gfx-unit-test/root-shader-parameter.cpp

```diff
@@ -0,0 +1,135 @@
+            ResourceState::CopySource);
+        bufferDesc.defaultState = ResourceState::UnorderedAccess;
+        bufferDesc.memoryType = MemoryType::DeviceLocal;
+
+        ComPtr<IBufferResource> numbersBuffer;
+        GFX_CHECK_CALL_ABORT(
+            device->createBufferResource(bufferDesc, (void*)&content, buffer.writeRef()));
+
+        return buffer;
+    }
+    void rootShaderParameterTestImpl(IDevice* device, UnitTestContext* context)
+    {
+        Slang::ComPtr<ITransientResourceHeap> transientHeap;
+        ITransientResourceHeap::Desc transientHeapDesc = {};
+        transientHeapDesc.constantBufferSize = 4096;
+        GFX_CHECK_CALL_ABORT(
+            device->createTransientResourceHeap(transientHeapDesc, transientHeap.writeRef()));
+
+        ComPtr<IShaderProgram> shaderProgram;
+        slang::ProgramLayout* slangReflection;
+        GFX_CHECK_CALL_ABORT(loadComputeProgram(device, shaderProgram, "root-shader-parameter", "computeMain", slangReflection));
+
+        ComputePipelineStateDesc pipelineDesc = {};
+        pipelineDesc.program = shaderProgram.get();
+        ComPtr<gfx::IPipelineState> pipelineState;
+        GFX_CHECK_CALL_ABORT(
+            device->createComputePipelineState(pipelineDesc, pipelineState.writeRef()));
+
+        Slang::List<ComPtr<IBufferResource>> buffers;
+        Slang::List<ComPtr<IResourceView>> srvs, uavs;
+
+        for (uint32_t i = 0; i < 9; i++)
+        {
+            buffers.add(createBuffer(device, i == 0 ? 10 : i));
+
+            ComPtr<IResourceView> bufferView;
+            IResourceView::Desc viewDesc = {};
+            viewDesc.type = IResourceView::Type::UnorderedAccess;
+            viewDesc.format = Format::Unknown;
+            GFX_CHECK_CALL_ABORT(
+                device->createBufferView(buffers[i], nullptr, viewDesc, bufferView.writeRef()));
+            uavs.add(bufferView);
+
+            viewDesc.type = IResourceView::Type::ShaderResource;
+            viewDesc.format = Format::Unknown;
+            GFX_CHECK_CALL_ABORT(
+                device->createBufferView(buffers[i], nullptr, viewDesc, bufferView.writeRef()));
+            srvs.add(bufferView);
+        }
+
+        ComPtr<IShaderObject> rootObject;
+        device->createMutableRootShaderObject(shaderProgram, rootObject.writeRef());
+
+        ComPtr<IShaderObject> g, s1, s2;
+        device->createMutableShaderObject(
+            slangReflection->findTypeByName("S0"), ShaderObjectContainerType::None, g.writeRef());
+        device->createMutableShaderObject(
+            slangReflection->findTypeByName("S1"), ShaderObjectContainerType::None, s1.writeRef());
+        device->createMutableShaderObject(
+            slangReflection->findTypeByName("S1"), ShaderObjectContainerType::None, s2.writeRef());
+
+        {
+            auto cursor = ShaderCursor(s1);
+            cursor["c0"].setResource(srvs[2]);
+            cursor["c1"].setResource(uavs[3]);
+            cursor["c2"].setResource(srvs[4]);
```

Comment:
Unrelated/random: when it comes time to add convenience features, it seems important that the `IShaderObject` interface should support convenience methods that work with `ShaderCursor`*.

In general, it would be good if a call like `someObject->setResource(<whatever>, srv)` is equivalent to:

```
auto cursor = ShaderCursor(someObject);
cursor[<whatever>].setResource(srv);
```

At the very least, that should support simple strings and indices (it is debatable whether it should support more general "path" expressions like `"a.b.c[3]"`, but it probably makes sense to allow those for convenience).

With such a convenience feature, this code could just become:

```
s1->setResource(0, srvs[2]);
s1->setResource(1, uavs[3]);
s1->setResource(2, srvs[4]);
```

The use of field indices in place of string names is something we should support and possibly even encourage for performance-minded programmers using `gfx`, since it avoids any string-based lookup while still being 100% portable across API back-ends.